File: leetcode-java/src/main/java/com/wezhyn/learn/dp/UniquePaths.java
package com.wezhyn.learn.dp;

/**
 * dp 依赖与之前有限个解
 * <p>
 * f(m,n)=f(m-1,n)+f(m,n-1)
 * m=0||n==0时，f(m,n)=1
 *
 * @author wezhyn
 * @since 04.12.2020
 */
public class UniquePaths {


    public int uniquePaths(int m, int n) {
        int[][] result=new int[m][n];
        for (int mi=1; mi < m; mi++) {
            for (int ni=1; ni < n; ni++) {
                result[mi][ni]=calculate(result, mi - 1, ni) + calculate(result, mi, ni - 1);
            }
        }
        int ans=calculate(result, m - 1, n - 1);
        return ans < 0 ? 2000000000 : ans;
    }

    public int calculate(int[][] result, int m, int n) {
        if (m==0 || n==0) {
            return 1;
        }
        return result[m][n];
    }

}

File: leetcode-kotlin/src/main/kotlin/com/wezhyn/learn/daily/UniquePaths.kt
package com.wezhyn.learn.daily

/**
 *
 * @author wezhyn
 * @since 08.16.2020
 *
 */
class UniquePaths {
    companion object {
        private val offset = intArrayOf(1, 0, 1)
    }

    fun uniquePaths(m: Int, n: Int): Int {
        val dp = Array(m) { IntArray(n) }
        dp[m - 1][n - 1] = 1

        for (i in m - 1 downTo 0) {
            for (j in n - 1 downTo 0) {
                for (k in 0..1) {
                    val x = i + offset[k]
                    val y = j + offset[k + 1]
                    if (x >= m || y >= n) {
                        continue
                    }
                    dp[i][j] += dp[x][y]
                }
            }
        }
        return if (dp[0][0] < 0) 2000000000 else dp[0][0]
    }
}